doc/persistence.adoc

```diff
@@ -0,0 +1,165 @@
+
+* booleans
+* numbers
+* strings
+* tables (objects/arrays)
+
+NOTE: Negative and positive infinity are represented by numbers which exceed the double number range by a lot (plus/minus `1e9999`).
+
+WARNING: The empty table `{}` (ambiguous: might be either `{}` or `[]`) and `nan` aren't supported by JSON and will be turned into `null`.
+
+Tables may either:
+. Contain only positive integer keys (represented as array) or
+. Contain only string keys (represented as object/dictionary)
+obviously, plenty of Lua tables aren't representable this way (boolean keys, mixed hash/list keys, fractional keys, inf keys)
+
+==== `minetest.write_json(data, [styled])`
+
+If the data is serializable as outlined above:: Returns a JSON string representing `data`. Formats the JSON nicely to improve readability if `styled` is truthy.
+Else:: Returns `nil` and one of the following errors if it fails:
+* `Can't use indexes with fractional part in JSON`
+* `Lua key to convert to JSON is not a string or number`
+* `Can't mix array and object values in JSON`
+
+WARNING: Hash tables containing only positive integer keys - a table of `minetest.hash_node_position` hashes for instance - will be turned into *an array with holes*, that is, all `nil` values from `1` to the maximum index will be filled with `null`, which means terrible performance and possibly very large output generated from very small input. *Do not allow direct access to `minetest.write_json`* as it can be trivially used to DoS your server.
+
+TIP: Sparse hash maps, should always be converted to objects, using only string keys, for JSON. You might do this just for serialization.
+
+WARNING: The JSON serializer is furthermore limited by its recursion depth: Only a recursion depth up to *16* is allowed. Very nested data structures can't be (de)serialized. Circular references will cause the JSON serializer to recurse until the maximum recursion depth is exceeded.
+
+TIP: Use the `json = assert(minetest.write_json(data))` idiom to not silently ignore errors when serializing.
+
+==== `minetest.parse_json(json, [nullvalue])`
+
+If `json` is valid JSON:: Returns the Lua value represented by the JSON string `json`, with `null` values deserialized to `nullvalue` (which defaults to `nil`).
+Else:: Returns `nil` and logs an error.
+
+TIP: If you must use JSON (to interact with a web interface, for instance), consider using your JSON library of choice; Lua-only implementations are available.
+
+=== Lua
+
+Minetest alternatively offers a serializer implemented in Lua, which serializes to Lua source code, available as  and `minetest.deserialize`. The following values are supported:
+
+* `nil`
+* booleans
+* numbers
+* strings
+* tables, including circular references
+* functions: *not recommended*, require mod security to be disabled
+** Lua-implementation specific bytecode: Functions serialized by PUC Lua won't work on LuaJIT and vice versa
+** Uses `string.dump` internally: Functions may not have upvalues
```

Comment:
This isn't super clear on what kind of function serialization is supported. Both bytecode and non-bytecode versions of functions can be deserialized (assuming the safe flag is false), but only non-bytecode functions can be serialized (iirc).